### PR TITLE
[Driver][SYCL] Improve handling of linker inputs for static lib proce…

### PR DESCRIPTION
**Pull request by mdtoguchi**
…ssing

When processing the command line for static archives, we go through the
direct to linker values (-Wl, -Xlinker) to find any additional libraries
and objects that should be included.  The trigger options for that scanning
was too broad, allowing too many options (including options with separate
arguments) to be incorrectly processed as objects and/or libraries.

Signed-off-by: Michael D Toguchi <michael.d.toguchi@intel.com>

### CODE REVIEW

File: clang/test/Driver/sycl-offload-static-lib-2.cpp [hunk 1]
```diff
@@ -90,11 +90,11 @@
...
 // WHOLE_STATIC_LIB_2: ld{{.*}} "[[INPUTO]]" "@[[ARGFILE]]"
 
-/// test -Wl,<arg> behaviors for special case handling of -z and -rpath
-// RUN: %clangxx -target x86_64-unknown-linux-gnu -fsycl -L/dummy/dir %t.o -Wl,-rpath,nopass -Wl,-z,nopass %t.a %t_2.a -### 2>&1 \
+/// test behaviors for special case handling of -z and -rpath
+// RUN: %clangxx -target x86_64-unknown-linux-gnu -fsycl -z anystring -L/dummy/dir %t.o -Wl,-rpath,nopass -Wl,-z,nopass %t.a %t_2.a -### 2>&1 \
 // RUN:   | FileCheck %s -check-prefixes=WL_CHECK
-// WL_CHECK-NOT: ld{{(.exe)?}}" "-r" {{.*}} "{{.*}}crt1.o" "{{.*}}crti.o" "-L/dummy/dir" {{.*}} "nopass" {{.*}} "{{.*}}crtn.o"
-// WL_CHECK: ld{{.*}}" "-rpath" "nopass" "-z" "nopass"
+// WL_CHECK-NOT: ld{{(.exe)?}} "-r" {{.*}} "{{.*}}crt1.o" "{{.*}}crti.o" "-L/dummy/dir" {{.*}}"anystring" {{.*}}"nopass" {{.*}} "{{.*}}crtn.o"
```
[the remark this answers is not in the document]
Comment (mdtoguchi) [reply clang/test/Driver/sycl-offload-static-lib-2.cpp]: I've updated the negative check - much more succinct now.

[the remark this answers is not in the document]
Comment (AGindinson) [reply clang/test/Driver/sycl-offload-static-lib-2.cpp]:
```
ld{{.*}} "-z" "anystring" {{.*}} "-rpath" "nopass" "-z" "nopass"
```
I'm still confused: how come this positive check not filtered by the presumably-negative check? `"-z"` should probably fit the "accept all" regex.

[the remark this answers is not in the document]
Comment (mdtoguchi) [reply clang/test/Driver/sycl-offload-static-lib-2.cpp]: Oh, I think I see the confusion.  The negative check is `ld -r` and the positive is for `ld -z`

[the remark this answers is not in the document]
Comment (AGindinson) [reply clang/test/Driver/sycl-offload-static-lib-2.cpp]: Ah, I'm finally getting it now :D The negative is for the partial link step, which I _chose to ignore_. Thanks for clarifying this!

**Pull request merged**